starter-code/src/functions-and-arrays.js

```diff
@@ -98,3 +252,51 @@ var matrix = [
   [20, 73, 35, 29, 78, 31, 90, 1, 74, 31, 49, 71, 48, 86, 81, 16, 23, 57, 5, 54],
   [1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52, 1, 89, 19, 67, 48]
 ];
+
+/**
+ * calculates he greatest product of four adjacent numbers in the same direction (up, down, left, right)
+ * @param matrix
+ */
+
+function greatestProduct(matrix){
+
+  var maxProduct = 0;
+  var product;
+
+  for(var i=0; i<matrix.length; i++){
+    var row = matrix[i];
+    for(var j=0; j<row.length; j++){
+
+      // up
+      product = 1;
+      for(var k=0; k<= Math.min(3, i); k++){
+        product *= matrix[i-k][j];
+      }
+      maxProduct = (maxProduct<product)?product:maxProduct;
+
+      // down
+      product = 1;
+      for(var k=0; k<= Math.min(3, matrix.length - i - 1); k++){
+        console.log(i + " " +j + " " + k);
+        product *= matrix[i+k][j];
+      }
+      maxProduct = (maxProduct<product)?product:maxProduct;
+
+      // right
+      product = 1;
+      for(var k=0; k<= Math.min(3, row.length - j - 1); k++){
+        product *= matrix[i][j+k];
+      }
+      maxProduct = (maxProduct<product)?product:maxProduct;
+
+      // left
+      product = 1;
+      for(var k=0; k<= Math.min(3, j); k++){
+        product *= matrix[i][j-k];
+      }
+      maxProduct = (maxProduct<product)?product:maxProduct;
+    }
+  }
+
+  return maxProduct
```

Comment:
Nothing to say! the code is clean, you seem to be very advanced in your Javascript knowledge so I can only encourage you to keep up the good work!

Reply:
If you want to level up your javascript skills and logic, come and talk to us, we'll give you some games :)